### PR TITLE
Abstract out material settings

### PR DESCRIPTION
**Pull request by nyalldawson**
Opens the possibilities for non-phong based materials for 3d symbols

### CODE REVIEW

**Comment by nyalldawson** [comment]
@wonder-sk another one for you :)

File: q [hunk 1]
```diff
@@ -0,0 +1,521 @@
+[1mdiff --git a/src/core/layertree/qgslayertreelayer.cpp b/src/core/layertree/qgslayertreelayer.cpp[m
```
Comment (wonder-sk) [inline q:1]: Probably unintended for check in :-)

**Review by wonder-sk — COMMENTED** [review]
I really like where this work is going!

We just need to be careful about the kind of data on which a particular material is applied. For example, we use phong material settings class for all 3D symbols now, but it is a bit of cheating there anyway... because we support different things with different symbols, and when starting to use abstract base class, we should be probably more explicit about these differences. To summarize what we have:
- polygons - use phong material, expecting triangles
- buffered lines - use phong material, expecting triangles
- simple lines - use custom line material, expecting line data
- instanced points - use custom phong material capable of instancing, expecting triangles + points
- billboard points - use custom material, expecting points

We probably need to define rendering techniques, and any materials would register to a particular technique? So for example, for Phong shading we would have two materials - one from Qt for polygons and buffered lines, another one would be our custom for instanced rendering...

File: q (hunk 1 above)
Comment (nyalldawson) [reply q:1]: Oops, yep!

**Comment by nyalldawson** [comment]
> We just need to be careful about the kind of data on which a particular material is applied. For example, we use phong material settings class for all 3D symbols now, but it is a bit of cheating there anyway... because we support different things with different symbols, and when starting to use abstract base class, we should be probably more explicit about these differences. To summarize what we have:

> polygons - use phong material, expecting triangles
buffered lines - use phong material, expecting triangles

FYI - something has broken this in recent builds -- buffered lines now just show corrupted. I tried to track down what's happening, but can't work it out.

> simple lines - use custom line material, expecting line data
instanced points - use custom phong material capable of instancing, expecting triangles + points
billboard points - use custom material, expecting points

> We probably need to define rendering techniques, and any materials would register to a particular technique? So for example, for Phong shading we would have two materials - one from Qt for polygons and buffered lines, another one would be our custom for instanced rendering...

Yeah, that's a good point. Given that there's like to be a larger set of materials vs rendering techniques, I'm thinking of adding a flag to the abstract material class for "supported techniques". Then the symbol class itself can have a new "required techniques" method, which we can use at run time to restrict the choice of available materials to those which are compatible for the symbol. What do you think?

**Comment by nyalldawson** [comment]
> We probably need to define rendering techniques, and any materials would register to a particular technique? So for example, for Phong shading we would have two materials - one from Qt for polygons and buffered lines, another one would be our custom for instanced rendering...

@wonder-sk I'll address this part in the next PR (after #37952 is merged)

**Pull request merged**